src/libraries/System.IO/tests/Stream/Stream.ReadAtLeast.cs

```diff
@@ -0,0 +1,274 @@
+            for (int i = 20; i < 30; i++) Assert.Equal(0, buffer[i]);
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task ReadMoreThanMinimumBytes(bool async)
+        {
+            int readInvokedCount = 0;
+            var s = new DelegateStream(
+                canReadFunc: () => true,
+                readFunc: (array, offset, count) =>
+                {
+                    readInvokedCount++;
+
+                    int byteCount = Math.Min(count, 10);
+                    for (int i = 0; i < byteCount; i++) array[offset + i] = (byte)i;
+                    return byteCount;
+                });
+
+            // first try with a buffer that doesn't fill 3 full pages
+            byte[] buffer = new byte[28];
+
+            Assert.Equal(28, async ? await s.ReadAtLeastAsync(buffer, 22) : s.ReadAtLeast(buffer, 22));
+            Assert.Equal(3, readInvokedCount);
+            for (int i = 0; i < 10; i++) Assert.Equal(i, buffer[i]);
+            for (int i = 10; i < 20; i++) Assert.Equal(i - 10, buffer[i]);
+            for (int i = 20; i < 28; i++) Assert.Equal(i - 20, buffer[i]);
+
+            // now try with a buffer that is bigger than 3 pages
+            readInvokedCount = 0;
+            buffer = new byte[32];
+
+            Assert.Equal(30, async ? await s.ReadAtLeastAsync(buffer, 22) : s.ReadAtLeast(buffer, 22));
+            Assert.Equal(3, readInvokedCount);
+            for (int i = 0; i < 10; i++) Assert.Equal(i, buffer[i]);
+            for (int i = 10; i < 20; i++) Assert.Equal(i - 10, buffer[i]);
+            for (int i = 20; i < 30; i++) Assert.Equal(i - 20, buffer[i]);
+            for (int i = 30; i < 32; i++) Assert.Equal(0, buffer[i]);
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task ReadAtLeastZero(bool async)
+        {
+            int readInvokedCount = 0;
+            var s = new DelegateStream(
+                canReadFunc: () => true,
+                readFunc: (array, offset, count) =>
+                {
+                    readInvokedCount++;
+
+                    int byteCount = Math.Min(count, 10);
+                    for (int i = 0; i < byteCount; i++) array[offset + i] = (byte)i;
+                    return byteCount;
+                });
+
+            byte[] buffer = new byte[20];
+
+            Assert.Equal(10, async ? await s.ReadAtLeastAsync(buffer, 0) : s.ReadAtLeast(buffer, 0));
+            Assert.Equal(1, readInvokedCount);
+            for (int i = 0; i < 10; i++) Assert.Equal(i, buffer[i]);
+            for (int i = 10; i < 20; i++) Assert.Equal(0, buffer[i]);
+
+            // now try with an empty buffer
+            readInvokedCount = 0;
+            byte[] emptyBuffer = Array.Empty<byte>();
+
+            // The buffer is empty, so the Stream can only return 0 from Read. This looks like an end-of-stream,
```

Comment:
Thoughts on this scenario? What should happen when an empty buffer is passed? Should we throw an ArgumentException instead? We could avoid the up front check if we only checked if the buffer was empty after Read returns `0`

Reply:
iirc Sockets allow the users to perform a blocking zero-byte read which returns when data becomes available for reading. But I don't think that these new methods should be used for such scenarios. Even exactly the opposite, they should not be used.

My personal preference would be to throw for empty buffers, wait for users feedback and eventually change it before we ship .NET 7 RTM based on feedback.

Reply:
I changed it to throw a better exception in these cases.